blog/_posts/2019-10-07-crash-course-kubernetes-overview.md

```diff
@@ -0,0 +1,72 @@
+
+## Kubernetes Architecture
+Generally speaking, Kubernetes is containers all the way down. You bundle your app in a container, and it's run on a pod (which is also a container, but you don't need to worry about that.)
+
+**A _pod_ is the basic unit of your app**; the unit that you scale, the unit that your networking talks to, etc.
+
+Pods talk to each other and to the outside world via services; **a _service_ defines a set of pods and rules for communicating with them, from inside or outside the cluster.**
+
+Pods run on nodes; **_nodes_ are machines (physical or virtual) managed by Kubernetes which run 1+ pods, and some other daemons that let Kubernetes communicate with them.** Everything on your cluster is managed by the **master node**, which is like any other node except that, instead of running pods, it instead runs all the machinery that lets Kubernetes store/detect/change state, lets `kubectl` talk to the cluster, etc.
+
+![A hand-wavey architecture diagram](/assets/images/crash-course-kubernetes-overview/k8s-arch.png)
+*An extremely high-level, hand-wave-y architecture diagram.*
+
+### Where does all this live?
+The short answer is: in virtual machines! Kubernetes nodes are generally all VMs, which may be located on your computer--when using local Kubernetes runtimes like [Minikube](https://github.com/kubernetes/minikube), [Kubernetes for Docker for Mac](https://www.docker.com/blog/docker-mac-kubernetes/), or [Kubernetes in Docker (KinD)](https://github.com/kubernetes-sigs/kind)--or in the cloud--when using runtimes like [Google Kubernetes Engine (GKE)](https://cloud.google.com/kubernetes-engine/) or [Amazon Elastic Kubernetes Service (Amazon EKS)](https://aws.amazon.com/eks/).
+
+The even shorter answer is: it doesn't matter! You as the user don't need to know the location of the VM(s) that you're talking to; all you need is the address of the Kubernetes API server, which lives on the master node and allows you (via `kubectl`) to communicate with the rest of your cluster. If your kubecontext knows where the API server lives, then it doesn't matter whether it's on a VM in your own computer or on a server in some far-flung AWS AZ -- just send it a command, and Kubernetes will take care of the rest.
+
+## Okay, why should I care?
+Well, you don't have to!  Everyone uses Kubernetes at a different level of abstraction; for instance, i operated quite happily for a while without understanding precisely how a `kubectl apply` command results in a new pod… until i needed to know, and then this knowledge gap came back to bite me. That won't even necessarily be the case for you--maybe you'll never _need_ this knowledge of the guts of Kubernetes--but it will probably help your debugging abilities if you have accurate mental models.
```

Comment:
```suggestion
Well, you don't have to!  Everyone uses Kubernetes at a different level of abstraction; for instance, I operated quite happily for a while without understanding precisely how a `kubectl apply` command results in a new pod… until i needed to know, and then this knowledge gap came back to bite me. That won't even necessarily be the case for you--maybe you'll never _need_ this knowledge of the guts of Kubernetes--but it will probably help your debugging abilities if you have accurate mental models.
```